### ISSUE TITLE
Combine

### ISSUE THREAD

**Issue by phuongddx** [phuongddx/data-structure#7]
`CurrentValueSubject` is a part of Apple's Combine framework, which is designed to work with Swift's asynchronous programming model. It is one of the `Subject` types provided by Combine, and it has unique characteristics that differentiate it from other `Subject` types in the framework. Here's a comparison of `CurrentValueSubject` with other subjects in Combine:

### `CurrentValueSubject`
- **Initial Value**: Requires an initial value upon creation.
- **Value Access**: Stores the most recent value and allows access to it via the `value` property.
- **Behavior**: When a new subscriber subscribes, it immediately receives the current value.
- **Use Case**: Useful when you need to hold and provide the latest value to new subscribers.

### `PassthroughSubject`
- **Initial Value**: Does not require an initial value.
- **Value Access**: Does not store any values; it simply passes through values it receives.
- **Behavior**: New subscribers do not receive any previously emitted values; they only receive values emitted after they subscribe.
- **Use Case**: Suitable for scenarios where you want to broadcast events or messages without maintaining state.

### `ReplaySubject` (Hypothetical, as Combine doesn't include this)
- **Initial Value**: Not applicable; typically allows buffering of a specified number of emitted values.
- **Value Access**: Buffers a specified number of the most recent values and replays them to new subscribers.
- **Behavior**: New subscribers receive a replay of the buffered values.
- **Use Case**: Ideal for situations where you need to provide new subscribers with a history of recent values.

### Comparison Table

| Feature                  | `CurrentValueSubject`   | `PassthroughSubject` | `ReplaySubject` (Hypothetical) |
|--------------------------|-------------------------|----------------------|-------------------------------|
| Requires Initial Value   | Yes                     | No                   | No                            |
| Stores Latest Value      | Yes                     | No                   | Yes                           |
| Immediate Value Delivery | Yes, delivers current value to new subscribers | No | Yes, replays buffered values to new subscribers |
| Use Case                 | Holding and providing the latest value | Broadcasting events/messages without state | Providing a history of recent values |

### Practical Example
**Using `CurrentValueSubject`**:
```swift
import Combine

let subject = CurrentValueSubject<Int, Never>(0)

// Subscriber 1
let subscriber1 = subject.sink { value in
    print("Subscriber 1 received: \(value)")
}

subject.value = 1

// Subscriber 2
let subscriber2 = subject.sink { value in
    print("Subscriber 2 received: \(value)")
}

subject.send(2)
```
**Output**:
```
Subscriber 1 received: 0
Subscriber 1 received: 1
Subscriber 2 received: 1
Subscriber 1 received: 2
Subscriber 2 received: 2
```

**Using `PassthroughSubject`**:
```swift
import Combine

let passthroughSubject = PassthroughSubject<Int, Never>()

// Subscriber 1
let subscriber1 = passthroughSubject.sink { value in
    print("Subscriber 1 received: \(value)")
}

passthroughSubject.send(1)

// Subscriber 2
let subscriber2 = passthroughSubject.sink { value in
    print("Subscriber 2 received: \(value)")
}

passthroughSubject.send(2)
```
**Output**:
```
Subscriber 1 received: 1
Subscriber 1 received: 2
Subscriber 2 received: 2
```

In summary, `CurrentValueSubject` is ideal when you need to maintain and provide the latest value to new subscribers, whereas `PassthroughSubject` is more suitable for broadcasting values without maintaining any state. The hypothetical `ReplaySubject` would be useful for scenarios requiring a history of emitted values.
